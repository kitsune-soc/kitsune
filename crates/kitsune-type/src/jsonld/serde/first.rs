use super::OptionSeed;
use core::{
    fmt::{self, Formatter},
    marker::PhantomData,
};
use serde::de::{
    self, Deserialize, DeserializeSeed, Deserializer, IgnoredAny, IntoDeserializer, SeqAccess,
};

/// Deserialises the first element of a JSON-LD set.
pub struct First<T> {
    seed: T,
}

struct Visitor<T>(T);

impl<'de, T> First<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    pub fn new() -> Self {
        Self::with_seed(PhantomData)
    }

    pub fn deserialize<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        Self::new().deserialize(deserializer)
    }
}

impl<'de, T> First<T>
where
    T: DeserializeSeed<'de>,
{
    pub fn with_seed(seed: T) -> Self {
        Self { seed }
    }
}

// XXX: Intentionally limiting to `First<PhantomData<_>>` rather than `First<_>` to help inference
// of the type parameter of `Optional::<First<_>>::deserialize`.
impl<'de, T> Default for First<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    fn default() -> Self {
        Self::new()
    }
}

impl<'de, T> DeserializeSeed<'de> for First<T>
where
    T: DeserializeSeed<'de>,
{
    type Value = T::Value;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(Visitor(self.seed))
    }
}

impl<'de, T> de::Visitor<'de> for Visitor<T>
where
    T: DeserializeSeed<'de>,
{
    type Value = T::Value;

    fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str(super::EXPECTING_SET)
    }

    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: SeqAccess<'de>,
    {
        let mut seed = OptionSeed(Some(self.0));
        let value = if let Some(value) = seq.next_element_seed(&mut seed)? {
            // Unwrapping is fine here because the first call to `OptionSeed::deserialize` always
            // returns a `Some` and `next_element_seed` can only call it at most once because its
            // signature takes the seed by value.
            let value = value.unwrap();
            while let Some(IgnoredAny) = seq.next_element()? {}
            value
        } else if let Some(seed) = seed.0 {
            seed.deserialize(().into_deserializer())?
        } else {
            // Weirdly, the `SeqAccess` has consumed the seed yet it didn't return a value.
            return Err(de::Error::invalid_length(0, &super::EXPECTING_SET));
        };

        Ok(value)
    }

    forward_to_into_deserializer! {
        fn visit_bool(bool);
        fn visit_i8(i8);
        fn visit_i16(i16);
        fn visit_i32(i32);
        fn visit_i64(i64);
        fn visit_i128(i128);
        fn visit_u8(u8);
        fn visit_u16(u16);
        fn visit_u32(u32);
        fn visit_u64(u64);
        fn visit_u128(u128);
        fn visit_f32(f32);
        fn visit_f64(f64);
        fn visit_char(char);
        fn visit_str(&str);
        fn visit_borrowed_str(&'de str);
        fn visit_string(String);
        fn visit_bytes(&[u8]);
        fn visit_borrowed_bytes(&'de [u8]);
        fn visit_byte_buf(Vec<u8>);
        fn visit_none();
        fn visit_some();
        fn visit_unit();
        fn visit_newtype_struct();
        fn visit_map();
        fn visit_enum();
    }
}

#[cfg(test)]
mod tests {
    use super::{super::into_deserializer, First};
    use core::marker::PhantomData;

    #[test]
    fn single() {
        let data = 42;
        assert_eq!(First::deserialize(into_deserializer(data)), Ok(data));
    }

    #[test]
    fn seq() {
        let data = vec![42, 21];
        assert_eq!(First::deserialize(into_deserializer(data)), Ok(42));
    }

    #[test]
    fn empty() {
        let data: Vec<u32> = Vec::new();
        assert_eq!(
            First::<PhantomData<Option<u32>>>::deserialize(into_deserializer(data)),
            Ok(None)
        );
    }
}
