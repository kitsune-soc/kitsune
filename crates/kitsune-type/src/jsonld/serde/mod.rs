//! Serde helpers to translate JSON-LD data structures.
//!
//! ## JSON-LD `@set`
//!
//! When a JSON-LD term's `@container` is unspecified or is set to `@set`, JSON entry values in the
//! following groups are semantically equivalent, respectively:
//!
//! - A non-array value (`"value"`) and a single-value array of the same value (`["value"]`)
//! - An empty array (`[]`), `null` and an absent entry
//!
//! The following helpers in the module deserialise a set as a sequence:
//!
//! - [`Set`]
//! - [`IdSet`]
//!
//! The following helpers deserialise a single value or `null` from a set:
//!
//! - [`First`]
//! - [`FirstId`]
//! - [`FirstOk`]
//!
//! ## JSON-LD `@id`
//!
//! When a JSON-LD term's `@type` is set to `@id`, a JSON entry value of a single (IRI) string
//! (`"http://example.com/"`) is a shorthand for a Linked Data node identified by that string
//! (`{"@id": "http://example.com/"}`.
//!
//! The following helpers deserialise the node identifier string(s):
//!
//! - [`Id`]
//! - [`FirstId`]
//! - [`IdSet`]

macro_rules! forward_to_into_deserializer {
    (
        fn visit_borrowed_str($T:ty);
        $($rest:tt)*
    ) => {
        fn visit_borrowed_str<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::value::BorrowedStrDeserializer::new(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_borrowed_bytes($T:ty);
        $($rest:tt)*
    ) => {
        fn visit_borrowed_bytes<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::value::BorrowedBytesDeserializer::new(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_none();
        $($rest:tt)*
    ) => {
        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(()))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_some();
        $($rest:tt)*
    ) => {
        fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            self.0.deserialize(deserializer)
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_unit();
        $($rest:tt)*
    ) => {
        fn visit_unit<E>(self) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(()))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_newtype_struct();
        $($rest:tt)*
    ) => {
        fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            self.0.deserialize(deserializer)
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_seq();
        $($rest:tt)*
    ) => {
        fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::SeqAccess<'de>,
        {
            self.0.deserialize(serde::de::value::SeqAccessDeserializer::new(seq))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_map();
        $($rest:tt)*
    ) => {
        fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::MapAccess<'de>,
        {
            self.0.deserialize(serde::de::value::MapAccessDeserializer::new(map))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_enum();
        $($rest:tt)*
    ) => {
        fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::EnumAccess<'de>,
        {
            self.0.deserialize(serde::de::value::EnumAccessDeserializer::new(data))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn $name:ident($T:ty);
        $($rest:tt)*
    ) => {
        fn $name<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    () => {};
}

mod first;
mod first_id;
mod first_ok;
mod id;
mod id_set;
mod optional;
mod set;

pub use self::first::First;
pub use self::first_id::FirstId;
pub use self::first_ok::FirstOk;
pub use self::id::Id;
pub use self::id_set::IdSet;
pub use self::optional::Optional;
pub use self::set::Set;

use core::{
    fmt::{self, Formatter},
    marker::PhantomData,
};
use serde::de::{
    self,
    value::{EnumAccessDeserializer, MapAccessDeserializer, SeqAccessDeserializer},
    DeserializeSeed, Deserializer, EnumAccess, IntoDeserializer, MapAccess, SeqAccess, Visitor,
};

const EXPECTING_SET: &str = "a JSON-LD set";

/// A wrapper to implement `IntoDeserializer` for an `impl Deserializer`, because Serde somehow
/// doesn't provide a blanket impl.
struct DeserializerIntoDeserializer<D>(D);

/// A `DeserializeSeed` that catches a recoverable error and returns it as a successful value.
struct CatchError<T, E> {
    seed: T,
    marker: PhantomData<fn() -> E>,
}

struct OptionSeed<T>(Option<T>);

impl<'de, D> IntoDeserializer<'de, D::Error> for DeserializerIntoDeserializer<D>
where
    D: Deserializer<'de>,
{
    type Deserializer = D;

    fn into_deserializer(self) -> Self::Deserializer {
        self.0
    }
}

impl<'de, T> DeserializeSeed<'de> for &mut OptionSeed<T>
where
    T: DeserializeSeed<'de>,
{
    type Value = Option<T::Value>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        if let Some(seed) = self.0.take() {
            seed.deserialize(deserializer).map(Some)
        } else {
            Ok(None)
        }
    }
}

impl<'de, T, E> CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    pub fn new(seed: T) -> Self {
        Self {
            seed,
            marker: PhantomData,
        }
    }
}

impl<'de, T, E> DeserializeSeed<'de> for CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    type Value = Result<T::Value, E>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(self)
    }
}

macro_rules! catch_error_forward_to_into_deserializer {
    ($(fn $name:ident($T:ty);)*) => {$(
        fn $name<E2>(self, v: $T) -> Result<Self::Value, E2> {
            // We can tell that the error isn't fatal to the deserialiser because it's originated
            // from the already deserialised value `$t` rather than the deserialiser.
            Ok(self.seed.deserialize(serde::de::IntoDeserializer::into_deserializer(v)))
        }
    )*};
}

impl<'de, T, E> Visitor<'de> for CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    type Value = Result<T::Value, E>;

    fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str("a value")
    }

    fn visit_borrowed_str<E2>(self, v: &'de str) -> Result<Self::Value, E2>
    where
        E: de::Error,
    {
        Ok(self
            .seed
            .deserialize(de::value::BorrowedStrDeserializer::new(v)))
    }

    fn visit_borrowed_bytes<E2>(self, v: &'de [u8]) -> Result<Self::Value, E2>
    where
        E: de::Error,
    {
        Ok(self
            .seed
            .deserialize(de::value::BorrowedBytesDeserializer::new(v)))
    }

    fn visit_none<E2>(self) -> Result<Self::Value, E2> {
        Ok(self.seed.deserialize(().into_deserializer()))
    }

    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        self.deserialize(deserializer)
    }

    fn visit_unit<E2>(self) -> Result<Self::Value, E2> {
        Ok(self.seed.deserialize(().into_deserializer()))
    }

    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        self.deserialize(deserializer)
    }

    // XXX: The following methods cannot determine whether an error is recoverable. While we might
    // be able to implement them _right_ way by hooking into the `*Access` trait implementations and
    // recursively applying `CatchError` in the `*_seed` method calls, that wouldn't be worth the
    // effort since we're currently not using `FirstOk` (the only user of `CatchError` as of now)
    // for these types, and as for `visit_seq`, JSON-LD doesn't support nested sets anyway.
    fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
    where
        A: SeqAccess<'de>,
    {
        self.seed
            .deserialize(SeqAccessDeserializer::new(seq))
            .map(Ok)
    }

    fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
    where
        A: MapAccess<'de>,
    {
        self.seed
            .deserialize(MapAccessDeserializer::new(map))
            .map(Ok)
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: EnumAccess<'de>,
    {
        self.seed
            .deserialize(EnumAccessDeserializer::new(data))
            .map(Ok)
    }

    catch_error_forward_to_into_deserializer! {
        fn visit_bool(bool);
        fn visit_i8(i8);
        fn visit_i16(i16);
        fn visit_i32(i32);
        fn visit_i64(i64);
        fn visit_i128(i128);
        fn visit_u8(u8);
        fn visit_u16(u16);
        fn visit_u32(u32);
        fn visit_u64(u64);
        fn visit_u128(u128);
        fn visit_f32(f32);
        fn visit_f64(f64);
        fn visit_char(char);
        fn visit_str(&str);
        fn visit_string(String);
        fn visit_bytes(&[u8]);
        fn visit_byte_buf(Vec<u8>);
    }
}

#[cfg(test)]
fn into_deserializer<'de, T>(value: T) -> T::Deserializer
where
    T: serde::de::IntoDeserializer<'de, serde::de::value::Error>,
{
    serde::de::IntoDeserializer::into_deserializer(value)
}

#[cfg(test)]
mod tests {
    use super::{First, FirstId, FirstOk, IdSet, Optional};
    use serde::Deserialize;

    /// Checks that the types work for some random real-world-ish use cases.
    #[test]
    fn integrate() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Type {
            Note,
        }

        #[derive(Debug, Deserialize, PartialEq)]
        #[serde(rename_all = "camelCase")]
        struct Object {
            id: String,
            #[serde(deserialize_with = "FirstOk::deserialize")]
            r#type: Type,
            #[serde(deserialize_with = "FirstId::deserialize")]
            attributed_to: String,
            #[serde(default)]
            #[serde(deserialize_with = "Optional::<First<_>>::deserialize")]
            summary: Option<String>,
            #[serde(default)]
            #[serde(deserialize_with = "Optional::<First<_>>::deserialize")]
            content: Option<String>,
            #[serde(default)]
            #[serde(deserialize_with = "IdSet::deserialize")]
            to: Vec<String>,
            #[serde(default)]
            #[serde(deserialize_with = "IdSet::deserialize")]
            cc: Vec<String>,
        }

        let object = br#"
        {
            "@context": "https://www.w3.org/ns/activitystreams",
            "id": "https://example.com/notes/1",
            "type": "Note",
            "attributedTo": "https://example.com/actors/1",
            "summary": "An ordinary Note",
            "content": "Hello, world!",
            "to": ["https://example.com/actors/2"],
            "cc": ["https://www.w3.org/ns/activitystreams#Public"]
        }
        "#;
        let expected = Object {
            id: "https://example.com/notes/1".to_owned(),
            r#type: Type::Note,
            attributed_to: "https://example.com/actors/1".to_owned(),
            summary: Some("An ordinary Note".to_owned()),
            content: Some("Hello, world!".to_owned()),
            to: vec!["https://example.com/actors/2".to_owned()],
            cc: vec!["https://www.w3.org/ns/activitystreams#Public".to_owned()],
        };
        assert_eq!(simd_json::from_slice(&mut object.to_owned()), Ok(expected));

        let object = br#"
        {
            "@context": "https://www.w3.org/ns/activitystreams",
            "id": "https://example.com/notes/1",
            "type": ["http://schema.org/CreativeWork", "Note"],
            "attributedTo": [
                {
                    "id": "https://example.com/actors/1",
                    "type": "Person"
                },
                "https://example.com/actors/2"
            ],
            "summary": "A quirky Note",
            "to": "https://example.com/actors/3"
        }
        "#;
        let expected = Object {
            id: "https://example.com/notes/1".to_owned(),
            // Multiple `type`s including unknown ones:
            r#type: Type::Note,
            // Multiple `attributedTo`s and an embedded node:
            attributed_to: "https://example.com/actors/1".to_owned(),
            summary: Some("A quirky Note".to_owned()),
            // Absent `Option` field:
            content: None,
            // Single-value set:
            to: vec!["https://example.com/actors/3".to_owned()],
            // Absent `serde(default)` field:
            cc: vec![],
        };
        assert_eq!(simd_json::from_slice(&mut object.to_owned()), Ok(expected));
    }
}
