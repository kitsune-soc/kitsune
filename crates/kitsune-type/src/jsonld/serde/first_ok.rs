use super::CatchError;
use core::{
    fmt::{self, Formatter},
    marker::PhantomData,
};
use serde::de::{
    self,
    value::{EnumAccessDeserializer, MapAccessDeserializer},
    Deserialize, DeserializeSeed, Deserializer, EnumAccess, IgnoredAny, IntoDeserializer,
    MapAccess, SeqAccess,
};

// XXX: Conceptually, we could decompose it into `First` and a helper type that filters successfully
// deserialised elements in a JSON-LD set. In practice, however, the latter type cannot be
// implemented (at least straightforwardly) because it would need to hook the
// `SeqAccess::next_element_seed` method, where we cannot clone the generic seed value like we're
// doing in `Visitor::visit_seq` below.

/// Deserialises a single element from a JSON-LD set.
///
/// It tries to deserialise each of the elements in the set and returns the first one successfully
/// deserialised.
///
/// The detection of recoverable errors is a "best effort" check and won't work for maps for
/// example, although it works for strings. It's suitable for tag-like fields like `"type"`.
pub struct FirstOk<T> {
    seed: T,
}

struct Visitor<T>(T);

impl<'de, T> FirstOk<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    pub fn new() -> Self {
        Self::with_seed(PhantomData)
    }

    pub fn deserialize<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        Self::new().deserialize(deserializer)
    }
}

impl<'de, T> FirstOk<T>
where
    T: DeserializeSeed<'de> + Clone,
{
    pub fn with_seed(seed: T) -> Self {
        Self { seed }
    }
}

impl<'de, T> Default for FirstOk<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    fn default() -> Self {
        Self::new()
    }
}

impl<'de, T> DeserializeSeed<'de> for FirstOk<T>
where
    T: DeserializeSeed<'de> + Clone,
{
    type Value = T::Value;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(Visitor(self.seed))
    }
}

macro_rules! forward_to_into_deserializer {
    ($(fn $name:ident($T:ty);)*) => {$(
        fn $name<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            self.0
                .clone()
                .deserialize(serde::de::IntoDeserializer::into_deserializer(v))
                // No (deserialisable) element in the (single-value) set.
                // Interpret it as equivalent to `null` according to the JSON-LD data model.
                .or_else(|_: E| self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(())))
        }
    )*};
}

impl<'de, T> de::Visitor<'de> for Visitor<T>
where
    T: DeserializeSeed<'de> + Clone,
{
    type Value = T::Value;

    fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str(super::EXPECTING_SET)
    }

    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: SeqAccess<'de>,
    {
        loop {
            match seq.next_element_seed(CatchError::<_, A::Error>::new(self.0.clone()))? {
                Some(Ok(value)) => {
                    while let Some(IgnoredAny) = seq.next_element()? {}
                    return Ok(value);
                }
                Some(Err(_)) => {}
                None => return self.0.deserialize(().into_deserializer()),
            }
        }
    }

    fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.0
            .clone()
            .deserialize(de::value::BorrowedStrDeserializer::new(v))
            .or_else(|_: E| self.0.deserialize(().into_deserializer()))
    }

    fn visit_borrowed_bytes<E>(self, v: &'de [u8]) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.0
            .clone()
            .deserialize(de::value::BorrowedBytesDeserializer::new(v))
            .or_else(|_: E| self.0.deserialize(().into_deserializer()))
    }

    fn visit_none<E>(self) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.0.deserialize(().into_deserializer())
    }

    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        FirstOk::with_seed(self.0).deserialize(deserializer)
    }

    fn visit_unit<E>(self) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        self.0.deserialize(().into_deserializer())
    }

    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        FirstOk::with_seed(self.0).deserialize(deserializer)
    }

    fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
    where
        A: MapAccess<'de>,
    {
        self.0.deserialize(MapAccessDeserializer::new(map))
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: EnumAccess<'de>,
    {
        self.0.deserialize(EnumAccessDeserializer::new(data))
    }

    forward_to_into_deserializer! {
        fn visit_bool(bool);
        fn visit_i8(i8);
        fn visit_i16(i16);
        fn visit_i32(i32);
        fn visit_i64(i64);
        fn visit_i128(i128);
        fn visit_u8(u8);
        fn visit_u16(u16);
        fn visit_u32(u32);
        fn visit_u64(u64);
        fn visit_u128(u128);
        fn visit_f32(f32);
        fn visit_f64(f64);
        fn visit_char(char);
        fn visit_str(&str);
        fn visit_string(String);
        fn visit_bytes(&[u8]);
        fn visit_byte_buf(Vec<u8>);
    }
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use super::super::into_deserializer;
    use super::FirstOk;

    #[test]
    fn simple() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Test {
            A,
        }
        let data = "A";
        assert_eq!(FirstOk::deserialize(into_deserializer(data)), Ok(Test::A));
    }

    #[test]
    fn simple_fail() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Test {
            A,
        }

        let data = "B";
        assert_eq!(
            FirstOk::deserialize(into_deserializer(data)),
            Ok(None::<Test>)
        );
    }

    #[test]
    fn seq() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Test {
            A,
            B,
        }

        let data = vec!["C", "B", "A"];
        assert_eq!(FirstOk::deserialize(into_deserializer(data)), Ok(Test::B));
    }

    #[test]
    fn seq_fail() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Test {
            A,
            B,
        }

        let data = vec!["C", "D"];
        assert_eq!(
            FirstOk::deserialize(into_deserializer(data)),
            Ok(None::<Test>)
        );
    }
}
