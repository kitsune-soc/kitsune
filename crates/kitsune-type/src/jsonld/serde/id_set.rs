use super::{Id, Set};
use core::marker::PhantomData;
use serde::de::{Deserialize, DeserializeSeed, Deserializer};

/// Deserialises a JSON-LD set of nodes as a sequence of node identifier strings.
pub struct IdSet<T> {
    seed: T,
}

impl<'de, T> IdSet<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    pub fn new() -> Self {
        Self::with_seed(PhantomData)
    }

    pub fn deserialize<D>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
    {
        Self::new().deserialize(deserializer)
    }
}

impl<'de, T> IdSet<T>
where
    T: DeserializeSeed<'de>,
{
    pub fn with_seed(seed: T) -> Self {
        Self { seed }
    }
}

impl<'de, T> Default for IdSet<PhantomData<T>>
where
    T: Deserialize<'de>,
{
    fn default() -> Self {
        Self::new()
    }
}

impl<'de, T> DeserializeSeed<'de> for IdSet<T>
where
    T: DeserializeSeed<'de>,
{
    type Value = T::Value;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        Set::with_seed(Id::with_seed(self.seed)).deserialize(deserializer)
    }
}

#[cfg(test)]
mod tests {
    use super::{super::into_deserializer, IdSet};
    use serde::Deserialize;
    use serde_test::{assert_de_tokens, Token};
    use std::collections::HashMap;

    #[test]
    fn single_string() {
        let data = "http://example.com/";
        assert_eq!(
            IdSet::deserialize(into_deserializer(data)),
            Ok(vec![data.to_owned()])
        );
    }

    #[test]
    fn single_embedded() {
        let data: HashMap<_, _> = [("id", "http://example.com/")].into_iter().collect();
        assert_eq!(
            IdSet::deserialize(into_deserializer(data)),
            Ok(vec!["http://example.com/".to_owned()])
        );
    }

    #[test]
    fn seq() {
        #[derive(Debug, Deserialize, PartialEq)]
        #[serde(transparent)]
        struct Test {
            #[serde(deserialize_with = "IdSet::deserialize")]
            term: Vec<String>,
        }

        assert_de_tokens(
            &Test {
                term: vec![
                    "http://example.com/1".to_owned(),
                    "http://example.com/2".to_owned(),
                ],
            },
            &[
                Token::Seq { len: Some(2) },
                Token::Str("http://example.com/1"),
                Token::Map { len: None },
                Token::Str("id"),
                Token::Str("http://example.com/2"),
                Token::MapEnd,
                Token::SeqEnd,
            ],
        );
    }
}
