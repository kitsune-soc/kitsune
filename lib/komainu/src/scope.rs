use compact_str::CompactString;
use indexmap::{set, IndexSet};
use serde::Deserialize;
use std::{
    convert::Infallible,
    fmt::{self, Display},
    str::FromStr,
};

#[derive(Clone, Debug, Default, Deserialize)]
#[serde(transparent)]
pub struct Scope {
    inner: IndexSet<CompactString>,
}

impl FromStr for Scope {
    type Err = Infallible;

    #[inline]
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(s.split_whitespace().collect())
    }
}

impl Scope {
    #[inline]
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    #[inline]
    pub fn insert<Item>(&mut self, item: Item)
    where
        Item: Into<CompactString>,
    {
        self.inner.insert(item.into());
    }

    /// Determine whether `self` can be accessed by `resource`
    ///
    /// This implies that `resource` is equal to or a superset of `self`
    #[inline]
    #[must_use]
    pub fn can_be_accessed_by(&self, resource: &Self) -> bool {
        resource.inner.is_superset(&self.inner)
    }

    /// Determine whether `self` is allowed to perform an action
    /// for which you at least need `resource` scope
    #[inline]
    #[must_use]
    pub fn can_perform(&self, resource: &Self) -> bool {
        self.inner.is_superset(&resource.inner)
    }

    #[inline]
    pub fn iter(&self) -> impl Iterator<Item = &str> {
        self.inner.iter().map(CompactString::as_str)
    }
}

impl Display for Scope {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for item in itertools::intersperse(self.iter(), " ") {
            f.write_str(item)?;
        }

        Ok(())
    }
}

impl<Item> FromIterator<Item> for Scope
where
    Item: Into<CompactString>,
{
    #[inline]
    fn from_iter<T: IntoIterator<Item = Item>>(iter: T) -> Self {
        iter.into_iter().fold(Scope::new(), |mut acc, item| {
            acc.insert(item.into());
            acc
        })
    }
}

impl IntoIterator for Scope {
    type Item = CompactString;
    type IntoIter = set::IntoIter<Self::Item>;

    #[inline]
    fn into_iter(self) -> Self::IntoIter {
        self.inner.into_iter()
    }
}
