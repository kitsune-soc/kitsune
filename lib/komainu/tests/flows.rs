use self::fixtures::Fixture;
use bytes::Bytes;
use http_body_util::Empty;
use komainu::scope::Scope;
use serde::Serialize;
use std::str::FromStr;

mod fixtures;

#[allow(clippy::unreadable_literal)]
const RNG_SEED: u64 = 0xBADD1E;

#[derive(Serialize)]
struct SerdeResponse {
    body: Option<String>,
    #[serde(with = "http_serde::header_map")]
    headers: http::HeaderMap,
    #[serde(with = "http_serde::status_code")]
    status: http::StatusCode,
}

impl From<http::Response<()>> for SerdeResponse {
    #[inline]
    fn from(value: http::Response<()>) -> Self {
        let (parts, _body) = value.into_parts();

        Self {
            body: None,
            headers: parts.headers,
            status: parts.status,
        }
    }
}

impl From<http::Response<Bytes>> for SerdeResponse {
    #[inline]
    fn from(value: http::Response<Bytes>) -> Self {
        let (parts, body) = value.into_parts();
        let body = String::from_utf8(body.to_vec()).unwrap();

        let mut response: Self = http::Response::from_parts(parts, ()).into();
        response.body = Some(body);
        response
    }
}

#[futures_test::test]
async fn success() {
    fastrand::seed(RNG_SEED);
    let fixture = Fixture::generate();

    let uri = http::Uri::builder()
        .scheme("http")
        .authority("komainu.example")
        .path_and_query("/oauth/authorize?response_type=code&client_id=client_1")
        .build()
        .unwrap();

    let req = http::Request::builder()
        .uri(uri)
        .body(Empty::<Bytes>::new())
        .unwrap();
    let req = komainu::Request::read_from(req).await.unwrap();

    let acceptor = {
        let handle = fixture.code_grant.extract_raw(&req).await.unwrap();
        handle
            .accept("user id".into(), &Scope::from_str("read").unwrap())
            .await
            .unwrap()
    };

    let deny = {
        let handle = fixture.code_grant.extract_raw(&req).await.unwrap();
        handle.deny()
    };

    insta::assert_json_snapshot!(SerdeResponse::from(acceptor.into_response()));
    insta::assert_json_snapshot!(SerdeResponse::from(deny));
}
