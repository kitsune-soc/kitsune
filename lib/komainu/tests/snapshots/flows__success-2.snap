---
source: lib/komainu/tests/flows.rs
expression: "SerdeResponse::from(deny)"
---
{
  "body": null,
  "headers": {
    "location": "http://client_1.example/?error=access_denied"
  },
  "status": 302
}
